polkadot/service/src/lib.rs

```diff
@@ -284,40 +291,58 @@ impl Service {
 			chain: client.clone(),
 			transaction_pool: transaction_pool_adapter,
 		};
+
 		let network = network::Service::new(network_params)?;
+		let barrier = ::std::sync::Arc::new(Barrier::new(2));
+
+		let thread = {
+			let client = client.clone();
+			let network = network.clone();
+			let txpool = transaction_pool.clone();
+			let exit = exit.clone();
+
+			let thread_barrier = barrier.clone();
+			thread::spawn(move || {
+				network.start_network();
+
+				thread_barrier.wait();
+				let mut core = Core::new().expect("tokio::Core could not be created");
+				let events = client.import_notification_stream().for_each(move |notification| {
+					network.on_block_imported(notification.hash, &notification.header);
+					prune_imported(&*client, &*txpool, notification.hash);
+
+					Ok(())
+				});
+
+				core.handle().spawn(events);
+				if let Err(e) = core.run(exit) {
+					debug!("Polkadot service event loop shutdown with {:?}", e);
+				}
+				debug!("Polkadot service shutdown");
+			})
+		};
+
+		// before returning, make sure the network is started. avoids a race
+		// between the drop killing notification listeners and the new notification
+		// stream being started.
+		barrier.wait();
```

Comment:
Is this really necessary? There's no event loop blocked on the network anymore.

Reply:
it's also nice to make sure the network is started before we begin the consensus process.